src/apps/chifra/pkg/pinlib/chunk/chunk.go

```diff
@@ -0,0 +1,313 @@
+package chunk
+
+// Fetching, unzipping, validating and saving both index and bloom chunks
+
+import (
+	"bytes"
+	"compress/gzip"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"runtime"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/TrueBlocks/trueblocks-core/src/apps/chifra/pkg/config"
+	"github.com/TrueBlocks/trueblocks-core/src/apps/chifra/pkg/pinlib/manifest"
+	"github.com/TrueBlocks/trueblocks-core/src/apps/chifra/pkg/sigintTrap"
+	"github.com/panjf2000/ants/v2"
+)
+
+// The size of worker pools for downloading and saving data
+var poolSize = runtime.NumCPU() * 2
+
+type ChunkType uint
+
+const (
+	BloomChunk ChunkType = iota
+	IndexChunk
+)
+
+type ProgressEvent uint
+
+const (
+	ProgressDownloading ProgressEvent = iota
+	ProgressUnzipping
+	ProgressValidating
+	ProgressDone
+	ProgressError
+	ProgressAllDone
+)
+
+type ChunkProgress struct {
+	Event   ProgressEvent
+	Message string
+	Pin     *manifest.PinDescriptor
+}
+
+// jobResult type is used to carry both downloaded data and some
+// metadata to decompressing/file writing function through a channel
+type jobResult struct {
+	fileName string
+	fileSize int64
+	contents io.Reader
+	Pin      *manifest.PinDescriptor
+}
+
+// fetchResult type make it easier to return both download content and
+// download size information (for validation purposes)
+type fetchResult struct {
+	body      io.ReadCloser
+	totalSize int64
+}
+
+// Downloads a chunk using HTTP
+func fetchChunk(url string) (*fetchResult, error) {
+	response, err := http.Get(url)
+	body := response.Body
+	if response.StatusCode != 200 {
+		return nil, fmt.Errorf("wrong status code: %d", response.StatusCode)
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	fileSize, err := strconv.ParseInt(response.Header.Get("Content-Length"), 10, 64)
+	if err != nil {
+		return nil, err
+	}
+
+	return &fetchResult{
+		body:      body,
+		totalSize: fileSize,
+	}, nil
+}
+
+// For each pin in pins, downloads, unzips and saves the chunk of type indicated by chunkType.
+// Progress is reported to progressChannel.
+func GetChunksFromRemote(pins []manifest.PinDescriptor, chunkType ChunkType, progressChannel chan<- *ChunkProgress) {
+	// Downloaded content will wait for saving in this channel
+	writeChannel := make(chan *jobResult, poolSize)
+	// Context lets us handle Ctrl-C easily
+	ctx, cancel := context.WithCancel(context.Background())
+	cacheLayout := &CacheLayout{}
+	cacheLayout.New(chunkType)
+	var downloadWg sync.WaitGroup
```

Comment:
My thinking was that download takes a different amount of time than writing to the disk. So I didn't want to block writing jobs by waiting for download to complete. This code downloads chunks and push them to a channel, not waiting for write. Write workers just take the data from the channel and write it in parallel whenever a worker is free.

Because downloading, unzipping and validating happens in-memory, only valid files will be written to the destination directory.